### ISSUE TITLE
Losing shared flag on remount with EL6, "ida_remove called for id=<X> which is not allocated."

### ISSUE THREAD

**Issue by jthiltges** [apptainer/singularity#4851]
### Version of Singularity:

singularity-3.4.2-1.1.osg34.el6

### Behavior

Singularity with underlay and a sandbox image on EL6 causes a kernel warning message `ida_remove called for id=1 which is not allocated.` for each bind mount. This behavior occasionally leads to a crash.

When a bind mount is remounted on EL6, the shared flag is lost, apparently due to a kernel bug fixed in 2.6.33 and beyond:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=7b43a79f32c0a05e7562043af98e25c05c89b18e

Singularity triggers the bug when it remounts a shared mountpoint.
https://github.com/sylabs/singularity/blob/v3.5.1/internal/pkg/runtime/engine/singularity/container_linux.go#L691

This corruption of kernel data can lead to ida_remove warnings and occasional system crashes.

### Steps to reproduce this behavior

Run a sandbox image on EL6 with underlay enabled:
```
singularity exec /cvmfs/singularity.opensciencegrid.org/opensciencegrid/osg-wn:3.3-el6 echo hello
```

dmesg output will show ida_remove warning messages.

The underlying issue can be reproduced with the command line:

```
# mkdir -p /demo/{src,dest1,dest2}
# unshare -m /bin/bash
# mount --make-shared /
# mount --bind /demo/src /demo/dest1
# grep dest1 /proc/self/mountinfo
39 28 253:0 /demo/src /demo/dest1 rw,relatime shared:1 - ext4 /dev/mapper/VolGroup-lv_root rw,seclabel,barrier=1,data=ordered
```

The /demo/dest1 mountpoint is shared. Now we remount it.

```
# mount -o remount /demo/dest1
# grep dest1 /proc/self/mountinfo
39 28 253:0 /demo/src /demo/dest1 rw,relatime - ext4 /dev/mapper/VolGroup-lv_root rw,seclabel,barrier=1,data=ordered
```

The /demo/dest1 mountpoint is no longer shared. If we add an additional bind mount, then exit the namespace, we'll get a kernel error.

```
# mount --bind /demo/dest1 /demo/dest2
# exit
exit
# dmesg
ida_remove called for id=1 which is not allocated.
```

### Possible solutions

1. Only do the remount if it is necessary. (Needed on EL6?)

1. Re-share the directory with the equivalent of `mount --make-shared` before doing further bind mounts.
   Calling mount(..., MS_SHARED), without setting the MS_REMOUNT flag, seems to restore the shared flag. If this is done before further bind mounts, I'm not seeing ida_remove errors when the namespace is removed.

### What OS/distro are you running

Scientific Linux release 6.10 (Carbon)
kernel 2.6.32-754.17.1.el6.x86_64

### How did you install Singularity

OSG packaging

**Comment by DrDaveD** [comment]
@cclerget 

John also reported this to Red Hat, but it's too late in the EL6 lifetime for them to consider fixing this.

**Comment by DrDaveD** [comment]
I can reproduce this with 3.5.1 on el6 with alpine:
```
$ singularity build --sandbox alpine-sandbox docker://alpine
$ singularity exec alpine-sandbox echo hello
$ sudo tail -1 /var/log/messages
Dec  9 12:39:20 dwdosgdev kernel: ida_remove called for id=22 which is not allocated.
```

**Comment by cclerget** [comment]
Hi @jthiltges , #4853 should fix the issue, if you can confirm it works as expected ? Thanks !

**Comment by jthiltges** [comment]
I tested a build from master (e4d837e) with your #4853 patch, and can confirm that it looks fixed: no more ida_remove messages.

Thanks very much for your help.

**Comment by cclerget** [comment]
Thanks ! Fix included for the next 3.5.2 release
